### PR TITLE
Refactor essential file migration; use CollectionManager 

### PR DESCRIPTION
**Pull request by dae**
By performing the essential migration inside the CollectionManager
queue, we ensure that it has exclusive access to the collection files
for the duration of the call. This should hold even for legacy code, as
CollectionHelper.getCol() blocks until the queue is free.

I found the current code somewhat hard to follow, due in large part to
its Java origins, so I've also tried to tidy it up a fair bit. I think one
of the main reasons scoped storage has stalled so long is that people were
a bit intimidated by the code (I was at least!), and I hope that by making
it a bit simpler, people will feel more comfortable jumping in and helping
to get it over the line.

Notable changes:

- The validation and copying steps were intermixed and spread through
multiple files, which made it hard to follow. The source/dest folders are
now determined/validated separately, and then the actual copying
is done in a CollectionManager method. The chain starts in deckpicker.kt:migrate(),
as before.
- The typealiases and newtypes have been removed, as I felt they made
the code harder to follow. The newtype also served to validate the folder
was a scoped storage folder, but that's now done in our validation routine.
Instead, the variable names have been renamed to be clearer, eg
 destinationDirectory: ScopedAnkiDroidDirectory -> scopedDestinationDirectory: Directory
- -wal are now copied if found. Usually they would get
removed when the collection is closed, but if there is a stale -wal file,
SQLite won't clean it up until a write action is performed on the DB, so it's
safer to include it.
- The corruption check before and after migration has been removed. For one,
it was only checking some referential integrity issues, and not actually checking
if the SQLite DB was corrupt. We could change the code to fix this, but I don't
see the point - we're already confirming the bytes of the copied files exactly match
the bytes of the source files, so the extra checks seem superfluous. Doing a full
integrity check on large collections is also quite slow, and risks hanging the
UI if legacy code happens to try access the collection.
- Removed the CollectionHelper code that was opening another instance of the backend.
- Removed the LockedCollection machinery, as CollectionManager takes its place.
- Removed the RetriableException code, as it only existed to deal with concurrent
access that can not happen with CollectionManager.

Tentatively closes https://github.com/ankidroid/Anki-Android/issues/13870. The first report looked like migration being
launched twice in parallel, and the second appears to have been caused by
StudyOptionsFragment refreshing the screen (perhaps in onResume),
which called withCol.

### CODE REVIEW

**Review by BrayanDSO — COMMENTED** [review]
AFAICS, this overall looks good and I agree with the simplification changes made. Tried to do a migration with the new fake media dev options and it worked fine.

> people were a bit intimidated by the code

I can relate to that.

---

The two issues I found were the pre-commit hook commit, which can be split into another PR if it blocks this one, and the tests failing on Windows (4 tries with fails in the 2 Windows runners in the same tests file) which can be ignored it they aren't an easy fix since time is of the essence here.

File: AnkiDroid/build.gradle [hunk 1]
```diff
@@ -234,11 +234,22 @@ amazon {
...
+        // Only perform the copy if the destination file doesn't already exist
+        if (!destFile.exists()) {
+            copy {
+                from srcFile
+                into destFile.parentFile
+                fileMode 0755
+            }
+        } else {
+            println 'Git hook already exists, skipping copy'
+        }
```
[the remark this answers is not in the document]
Comment (dae) [reply AnkiDroid/build.gradle:251]: Yeah, good point. I'll drop this commit when I next push, and define JAVA_HOME in my .zshrc instead.

**Comment by dae** [comment]
Setting up Android Studio in a Windows VM again was a pain (should have left the old image around from last time :sweat: ), but the test appears to fixed now.

**Review by BrayanDSO — APPROVED** [review]
Pretty good. Thanks a lot, Damien

**Review by lukstbit — APPROVED** [review]
Looked at the code and ran the migration, everything LGTM.
Thank you!
